extend aws_request4 function to allow passing custom headers

KingBrewer opened the pull request:
Addresses: https://github.com/alertlogic/erlcloud/issues/123

After this change new request may be implemented as:

``` erlang
request_impl(Config, Action, Params, ApiVersion) ->
    QParams = [{"Action", Action}, {"Version", ApiVersion}|Params],
    case erlcloud_aws:aws_request4(post, undefined,
                                   Config#aws_config.xxx_host,
                                   undefined, "/", QParams, "xxx",
                                   [{"Accept", "application/json"}], Config) of
    {ok, Response} ->
        {ok, jsx:decode(Response)};
    {error, Reason} ->
        {error, Reason}
    end.
```

KingBrewer committed:
extend aws_request4 function to allow passing custom headers

nalundgaard commented on src/erlcloud_aws.erl:
```diff
@@ -161,8 +165,8 @@ aws_request4_no_update(Method, Protocol, Host, Port, Path, Params, Service, #aws
-
-    aws_request_form(Method, Protocol, Host, Port, Path, Query, SignedHeaders, Config).
+    aws_request_form(Method, Protocol, Host, Port, Path, Query,
+                     [SignedHeaders | Headers], Config).
```
Why not just properly `++` here, and dispense the unnecessary `lists:flatten/1` below?

nalundgaard commented:
Other than my previous comment about style/improper treatment of header list, and that I strongly feel this should be made against erlcloud/erlcloud, this looks reasonable to me.

kkuzmin commented:
Hi Jakub,

Please note that this change still doesn't allow you to pass JSON body inside a request to AWS.
Does it work for you?

Based on the issue description after this change example request could be implemented like this:

```
request_impl(Config, Action, Params, ApiVersion) ->
    QParams = [{"Action", Action}, {"Version", ApiVersion}|Params],
    case erlcloud_aws:aws_request4(post, undefined,
                                   Config#aws_config.xxx_host,
                                   undefined, "/", QParams, "xxx",
                                   [{"Accept", "application/json"}, {"content-type", "application/x-amz-json-1.1"}], Config) of
    {ok, Response} ->
        {ok, jsx:decode(Response)};
    {error, Reason} ->
        {error, Reason}
    end.
```

This will lead to having two `content-type` headers in one HTTP request because one more is added inside `erlcloud_aws:aws_request_form()`.

KingBrewer commented:
@kkuzmin I kept the compatibility with current users and in these cases there's no JSON in the content. Nothing blocks us to proceed further and getting rid of `application/x-www-form-urlencoded` content type from `aws_request_form`. But I think it's the case for separate PR.

kkuzmin commented:
+1

KingBrewer commented:
@nalundgaard I addressed your comments. And I don't mind submitting it against erlcloud/erlcloud, however as I said my another PR (cloudsearch) will be dependent on this change. I need final decision :). @motobob could you share your thoughts about it?

motobob commented:
@KingBrewer given how diverged AL and upstream forks are in terms of SigV4 and my delaying big merge of AL into upstream it does not make sense to close this but please make a copy PR to upstream. 
this one 👍
